src/compiler/checker.ts

```diff
@@ -17122,6 +17099,27 @@ namespace ts {
                 }
             }
 
+            function getInitialOrAssignedType(flow: FlowAssignment, reference: Node) {
+                const node = flow.node;
+                if (flow.flags & FlowFlags.Cached) {
+                    const cached = flowAssignmentTypes[getNodeId(node)];

```

Comment:
Doesn't the cache key need to involve the `initialType` argument passed into `getFlowTypeOfReference`, as that is invoked with multiple different values and affects what we would cache here?

Reply:
No, because what we're caching here is the type of the expression assigned to the reference, not the control flow type that we compute by combining this assigned type with `declaredType` and/or `initialType`. Resolution of the assigned type in no way depends on what was passed to `getFlowTypeOfReference` (because we're immediately calling out of the `getFlowTypeOfReference` closure).